File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/resource/HomeController.java
package uk.gov.justice.digital.hmpps.oauth2server.resource;

import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.servlet.ModelAndView;
import uk.gov.justice.digital.hmpps.oauth2server.landing.LandingService;

import java.util.stream.Collectors;

@Controller
public class HomeController {
    private final LandingService landingService;

    public HomeController(final LandingService landingService) {
        this.landingService = landingService;
    }

    @GetMapping("/")
    public ModelAndView home(final Authentication authentication) {
        final var services = landingService.findAllServices();

        // create a list of services that the user can see
        final var allowedServices = services.stream().
                filter((s) -> s.getRoles().isEmpty() || authentication.getAuthorities().stream().anyMatch((a) -> s.getRoles().contains(a.getAuthority()))).
                collect(Collectors.toList());
        return new ModelAndView("landing", "services", allowedServices);
    }

    @GetMapping("/terms")
    public String terms() {
        return "terms";
    }

}

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/resource/HomeController.kt
package uk.gov.justice.digital.hmpps.oauth2server.resource

import org.springframework.security.core.Authentication
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.servlet.ModelAndView
import uk.gov.justice.digital.hmpps.oauth2server.security.AuthSource
import uk.gov.justice.digital.hmpps.oauth2server.security.AuthSource.azuread
import uk.gov.justice.digital.hmpps.oauth2server.security.UserDetailsImpl
import uk.gov.justice.digital.hmpps.oauth2server.service.AuthServicesService
import uk.gov.justice.digital.hmpps.oauth2server.service.UserContextService

@Controller
class HomeController(
  private val authServicesService: AuthServicesService,
  private val userContextService: UserContextService,
) {
  @GetMapping("/")
  fun home(authentication: Authentication): ModelAndView {
    // special case for azure users - grab all their accounts and combine roles from them too
    val azureUserDetails = authentication.principal as UserDetailsImpl
    val authSource = AuthSource.fromNullableString(azureUserDetails.authSource)
    val authorities = if (authSource == azuread) {
      val users = userContextService.discoverUsers(authentication.principal as UserDetailsImpl, setOf())
      users.flatMap { it.authorities }
    } else authentication.authorities

    val services = authServicesService.listEnabled(authorities)

    return ModelAndView("landing", "services", services)
  }

  @GetMapping("/terms")
  fun terms(): String = "terms"
}
